hera_filters/dspec.py

```diff
@@ -2120,6 +2120,215 @@ def fit_solution_matrix(weights, design_matrix, cache=None, hash_decimal=10, fit
+    
+    if opkey not in cache:
+        # Normalize input array to -1 <= xg <= 1 unless a range is provided 
+        if xmin is None:
+            xmin = x.min()
+        if xmax is None:
+            xmax = x.max()
+        
+        xg = np.copy(x)
+        xg = 2 * (xg - xmin) / (xmax - xmin) - 1
+
+        # Compute number of polynomials needed
+        if nterms is not None:
+            kmax = np.max(nterms)
+        else:
+            nwindows = []
+            for eigval, fw in zip(eigenval_cutoff, filter_half_widths):
+                c = fw * np.pi * (xmax - xmin)
+                nw = (
+                    2 * (xmax - xmin) * fw
+                    + 2
+                    / np.pi ** 2
+                    * np.log(100 * c / np.pi + 25)
+                    * np.log(5 / (eigval * (1 - eigval)))
+                    + 7
+                )
+                nwindows.append(nw)
+            kmax = np.round(np.max(nw)).astype(int)
+
+        # Compute associated Legendre polynomials
+        polynomials = _normalized_legendre(xg, kmax)
+        
+        amat, _nterms = [], []
+        for fn, (fw, fc) in enumerate(zip(filter_half_widths, filter_centers)):
+            # Compute eigenvectors
+            c = fw * np.pi * (xmax - xmin)
+            eigenvals, eigenvecs = _calculate_amat(kmax, c)
+
+            # Sort eigenvectors by largest eigenvalue
+            idx = np.argsort(eigenvals)
+            eigenvecs = eigenvecs[:, idx]
+
+            # Calculate PSWFs from eigenvectors and legendre polynomials
+            pswf_vectors = polynomials @ eigenvecs
+
+            # If number of terms is given, truncate pswf vectors
+            if nterms is not None:
+                pswf_vectors = pswf_vectors[:, : nterms[fn]]
+                _nterms.append(nterms[fn])
+
+            # If eigenval_cutoff is given, estimate pswf eigenvalues
+            if eigenval_cutoff is not None:
+                # Get legendre polynomial at P(0)
+                poly = _normalized_legendre(np.array(0), kmax)
+                
+                # Evaluate PSWF a x=0
+                midpoint = poly @ eigenvecs
+                
+                # Compute eigenvalues
+                neven = np.arange(0, kmax, 2)
+                eigvals = np.abs(np.sqrt(2) * eigenvecs[0, neven] / midpoint[neven])
+                eigvals = (eigvals / eigvals.max()) ** 2
+                nt = np.max(neven[eigvals > eigenval_cutoff[fn]])
+                
+                # Truncate pswf vectors
+                pswf_vectors = pswf_vectors[:, :nt]
+                _nterms.append(nt)
+
+            # Apply filter center to design matrix
+            amat.append(
+                pswf_vectors
+                * np.exp(2j * np.pi * xg[:, None] * fc / (2 * np.pi * (xmax - xmin)))
+            )
+
+        # Stack design matrices
+        if len(amat) > 1:
+            amat = np.hstack(amat)
```

Comment:
This needs coverage.

Reply:
All lines for the new functions should be covered now